scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+                self._ygas = np.random.ranf() * 2.0 - 1.0
+                self._sgas = self._xgas * self._xgas + self._ygas * self._ygas
+            root = np.sqrt(-2.0 / self._sgas * np.log(self._sgas))
+            self._ranbyx = self._xgas * root
+            self._ranbyy = self._ygas * root
+            retval = self._ranbyy
+            self._usey = 0
+        else:
+            retval = self._ranbyx
+            self._usey = 1
+        return retval
+
+    @property
+    def result(self):
+        """ The OptimizeResult """
+        res = OptimizeResult()
+        res.x = self._xmini
+        res.fun = self._fvalue
+        res.message = self._message
+        res.nit = self._step_record
+        return res
+
+def gensa(func, x0, bounds, niter=500, T=5230., visitparam=2.62,
+        acceptparam=-5.0, maxtime=3600, maxcall=1e7, args=()):
+    """
+    Find the global minimum of a function using the Generalized Simulated
+    Annealing algorithm
+
+    Parameters
+    ----------
+    func : callable ``f(x, *args)``
+        Function to be optimized.  ``args`` can be passed as an optional item
+        as argument (see below)
+    x0 : ndarray
+        Optional initial guess. If not given (None), a random one is computed
+        using bounds sequence
+    bounds: sequence
+        Bounds for variable. (min, max) pairs for each element in x,
+        defining the bounds on that parameter.
+    niter: integer, optional
+        The number of gensa iterations
+    T: float, optional
+        Initial value for temperature
+    visitparam: float, optional
+        Parameter for visiting distribution
+    acceptparam: float, optional
+        Parameter for acceptance distribution
+    maxtime: integer, optional
+        Time limit for running the algorithm in seconds
+    maxcall: integer, optional
+        Soft limit for the number of objective function calls. If the
+        algorithm is in the middle of a local search, this number will be
+        exceeded, the algorithm will stop just after the local search is
+        done.
+    args: Optional arguments to be passed to the objective function call.
+
+
+    Returns
+    -------
+    res : OptimizeResult
+        The optimization result represented as a ``OptimizeResult`` object.
+        Important attributes are: ``x`` the solution array, ``fun`` the value
+        of the function at the solution, and ``message`` which describes the
+        cause of the termination.
+        See `OptimizeResult` for a description of other attributes.
+
+    Notes
+    -----
+    GenSA is an implementation of the General Simulated Annealing algorithm
+    (GSA [2]_). This stochastic approach generalizes CSA (Classical Simulated
+    Annealing) and FSA (Fast Simulated Annealing) to find the neighborhood of
+    minima, then calls a local method (lbfgs) to find their exact value.
+    The algorithm was originally implemented in C++/R and is explained in
```

Comment:
I don't think there is a need to say it was originally implemented in C++/R. Just say the algorithm is explained in [5]. However, should I use reference 2 (first sentence) or 5 to find out more details?